layout: Fix warts in NestedNavRow and SwitchRow

chrisbobbe opened the pull request:
And (experimental) use `NestedNavRow` in place of new duplicated code at the bottom of `SubscriptionsScreen`, from 9d11e787fd10d5ef07d7b1d8e02127c922e189c7.

See https://github.com/zulip/zulip-mobile/pull/5443#issuecomment-1192014031 for rationale.

chrisbobbe commented:
Before / after:

<img width=400 src="https://user-images.githubusercontent.com/22248748/180330026-7c8f3baf-e9f0-4872-82b0-6111dd8ba2d3.png" /> <img width=400 src="https://user-images.githubusercontent.com/22248748/180329878-0c2ce464-4274-40c7-b4d4-c88547f6555a.png" />

Before / after:

<img width=400 src="https://user-images.githubusercontent.com/22248748/180329907-8ad981eb-0ad0-4aaa-b0a8-c82a2b7ac6a9.png" /> <img width=400 src="https://user-images.githubusercontent.com/22248748/180330588-714444fd-148e-4cd1-bde0-3cf263c89735.png" />

Before / after:

<img width=400 src="https://user-images.githubusercontent.com/22248748/180330814-c983fbe3-6a31-4f3c-a58b-cb488a1ebb02.png" /> <img width=400 src="https://user-images.githubusercontent.com/22248748/180330800-cbbac165-a691-4995-89a7-f9f9cf723559.png" />

chrisbobbe commented on src/common/SwitchRow.js:
```diff
@@ -17,10 +17,11 @@ type Props = $ReadOnly<{|
-    marginVertical: 8,
     textAlign: 'center',
     marginLeft: 8,
     marginRight: 16,
```
```
SwitchRow: Use container minHeight for height

We care that the touch target is at least 48px in height, and we
care that the row can expand to accommodate a larger system font
size than the standard. This seems like a transparent way to
accomplish both.
```

Hmm, I think I too lazily copy-pasted this reasoning from `NestedNavRow`. This component, `SwitchRow`, isn't a touch target across its whole surface; only the switch is a touch target.

gnprice replied:
Yeah, I guess the touch target here is the `Switch`.

That does extend to the full height of the row after this change, so the effect is the same.

Perhaps set the `height` of the Switch to 48? And then leave the container without a height-related prop?

chrisbobbe replied:
> That does extend to the full height of the row after this change, so the effect is the same.

The `styles.listItem` is applying top and bottom padding of 8px, so the switch isn't extending to the full height of the row. There's still row for 8px above the top of the switch, and for 8px below the bottom.

The inspector shows that on iOS the switch has height 31 and width 51. So in the vertical direction, my device is actually giving the row a computed height of 31 + 8 + 8 = 47.

To make this row be the same height as the others (48) with common device settings, I suppose we could bump that 47 up to 48 (and handle any discrepancy on Android) by allotting more space for the switch, yeah.

chrisbobbe replied:
Or I could apply minHeight: 48 to the row, and replace the reasoning with something like

> To make this row be the same height as the others (48) with common device settings

rather than comments on how big touch targets should be.

gnprice replied:
> The `styles.listItem` is applying top and bottom padding of 8px, so the switch isn't extending to the full height of the row. There's still row for 8px above the top of the switch, and for 8px below the bottom.

Hmm, I see.

Does that padding end up as part of the touch target? I think we want it to be. In that case, setting the `minHeight` of the row to 48px makes sense.

If the padding isn't part of the touch target, then since 32px isn't tall enough for the touch target we'll want to make the touch target taller.

alya commented:
@chrisbobbe could you explain a bit what I should focus on for reviewing this PR? Thanks!

chrisbobbe commented:
(Alya, Greg, and I discussed this in the office, and IIRC we liked these changes. The next step is a code review from Greg.)

gnprice commented:
Thanks @chrisbobbe for these cleanups!

All looks good, except the small point you mentioned (and I just replied to) above.

chrisbobbe commented:
Thanks for the review! Revision pushed, addressing your comment with the approach I mentioned at https://github.com/zulip/zulip-mobile/pull/5446#discussion_r957914779.

chrisbobbe commented:
From discussion in the office today, unfortunately we don't have a nice, non-janky way to make the touch target extend up and down, such that if you tapped in the area between two switches, it would be taken as an interaction with one of them. From experimentation, the latter behavior is what you get on Android and iOS in their system settings UIs.

Ah, well; this should be ready for another code review.

gnprice commented:
Thanks! Ah well, indeed.

Looks good given that limitation of RN -- merging.
